src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java

```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+        }
+        AtomicInteger count = executionsMap.computeIfAbsent(tenantId, k -> new AtomicInteger(0));
+        if (count.incrementAndGet() <= maxExecutions) {
+            return true;
+        } else {
+            count.decrementAndGet(); // Rollback the increment
+            return false;
+        }
+    }
+
+    /**
+     * Releases an execution slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the execution slot.
+     * @param executionsMap The map tracking the number of active executions per tenant.
+     */
+    private static void release(String tenantId, ConcurrentHashMap<String, AtomicInteger> executionsMap) {
+        if (tenantId == null) {
+            return; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.get(tenantId);
+        if (count != null) {
+            count.decrementAndGet();
+        }
+    }
+
+    /**
+     * Create an action listener that releases provision throttle on failure only
+     * @param tenantId The tenant ID
+     * @param delegate the wrapped listener
+     * @return a listener wrapping the delegate that calls {@link #releaseProvision(String)} on failure of the wrapped listener.
+     */
+    public static ActionListener<WorkflowResponse> releaseProvisionOnFailureListener(
+        String tenantId,
+        ActionListener<WorkflowResponse> delegate
+    ) {
+        return ActionListener.notifyOnce(new ActionListener<>() {
+            @Override
+            public void onResponse(WorkflowResponse response) {
+                delegate.onResponse(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                try {
+                    releaseProvision(tenantId);
+                } finally {
+                    delegate.onFailure(e);
+                }
+            }
+        });
+    }
+
+    /**
+     * Create an action listener that releases deprovision throttle
+     * @param tenantId The tenant ID
+     * @param delegate the wrapped listener
+     * @return a listener wrapping the delegate that calls {@link #releaseDeprovision(String)} on the wrapped listener.
+     */
+    public static ActionListener<WorkflowResponse> releaseDeprovisionListener(String tenantId, ActionListener<WorkflowResponse> delegate) {
+        return ActionListener.notifyOnce(new ActionListener<>() {
+            @Override
+            public void onResponse(WorkflowResponse response) {
+                try {
+                    releaseDeprovision(tenantId);
+                } finally {
+                    delegate.onResponse(response);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                try {
+                    releaseDeprovision(tenantId);
+                } finally {
+                    delegate.onFailure(e);
+                }
+            }
+        });
```

Comment:
Can use runAfter to reduce the method completely
```java
return ActionListener.notifyOnce(ActionListener.runAfter(delegate, () -> releaseDeprovision(tenantId)));
```

Reply:
In theory, yes, we could use `runBefore()` or `runAfter()` here because we are calling `releaseDeprovision()` in both cases.  In fact, the code here is exactly what happens on `runBefore()`.  

However, this is just a lucky break because we use synchronous code for deprovisioning, so our action listener handling is simpler.  We can _not_ do it for the provision and reprovision code because we only call `releaseProvision()` on failure and _not_ on success.

In which case I'd prefer to keep this method implementation similar to the adjacent method in the same class.